develop/tutorials/articles/03-developing-a-web-application/04-writing-web-application-liferay/01-writing-a-web-application.markdown

```diff
@@ -0,0 +1,635 @@
+                throws PortletException, IOException {
+
+            PortletPreferences prefs = renderRequest.getPreferences();
+            String[] guestbookEntries = prefs.getValues("guestbook-entries",
+                    new String[1]);
+
+            if (guestbookEntries != null) {
+                List<Entry> entries = parseEntries(guestbookEntries);
+
+                renderRequest.setAttribute("entries", entries);
+            }
+
+            super.render(renderRequest, renderResponse);
+        }
+
+    This method retrieves the array of guestbook entries from portlet
+    preferences, converts it to a `List` of `Entry` objects, and places that
+    `List` into the request object. It then calls the parent class's `render`
+    method.
+
+2. Beneath the `render` method, add the following method that converts the
+   array to a `List` of your model objects:
+
+        private List<Entry> parseEntries(String[] guestbookEntries) {
+            List<Entry> entries = new ArrayList<Entry>();
+
+            for (String entry : guestbookEntries) {
+                String[] parts = entry.split("\\^", 2);
+                Entry gbEntry = new Entry(parts[0], parts[1]);
+                entries.add(gbEntry);
+            }
+
+            return entries;
+        }
+
+As you can see, this method splits the entries in the `String` array into two
+parts based on the caret (`^`) character.
+
+Now that you have your controller preparing your data for display, your next
+step is to implement the view, so users can see guestbook entries.
+
+### Displaying Guestbook Entries
+
+Liferay's development framework makes it easy to loop through data and display
+it nicely to the end user. You'll use a Liferay UI construct called the *Search
+Container* to make this happen.
+
+1. Add the these tags to your `view.jsp` below the closing
+   `</portlet:renderURL>` tag but before the opening `<aui:button-row>` tag:
+
+        <jsp:useBean id="entries" class="java.util.ArrayList" scope="request"/>
+
+        <liferay-ui:search-container>
+            <liferay-ui:search-container-results results="<%= entries %>" />
+
+            <liferay-ui:search-container-row
+                className="com.liferay.docs.guestbook.model.Entry"
+                modelVar="entry"
+            >
+                <liferay-ui:search-container-column-text property="message" />
+
+                <liferay-ui:search-container-column-text property="name" />
+            </liferay-ui:search-container-row>
+
+            <liferay-ui:search-iterator />
+        </liferay-ui:search-container>
+
+Save your work, deploy your application, and try adding some guestbook entries.
+
+## Next Steps
+
+Here are more tutorials that you might find interesting:
+
+[Creating Liferay Applications](/develop/tutorials/-/knowledge_base/7-0/creating-liferay-applications)
+
+[Creating Liferay Components](/develop/tutorials/-/knowledge_base/7-0/creating-liferay-components)
+
```

Comment:
This was really great! Much improved from the previous iteration. We're getting there!